crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -131,46 +131,97 @@ impl CommandQueue {
+                let meta = unsafe { cursor.cast::<CommandMeta>().read_unaligned() };
+                // Advance to the bytes just after `meta`, which represent a type-erased command.
+                // SAFETY: For most types of `Command`, the pointer immediately following the metadata
+                // is guaranteed to be in bounds. If the command is a zero-sized type (ZST), then the cursor
+                // might be 1 byte past the end of the buffer, which is safe.
+                cursor = unsafe { cursor.add(std::mem::size_of::<CommandMeta>()) };
+                // Construct an owned pointer to the command.
+                // SAFETY: It is safe to transfer ownership out of `self.bytes`, since the call to `set_len(0)` above
+                // guarantees that nothing stored in the buffer will get observed after this function ends.
+                // `cmd` points to a valid address of a stored command, so it must be non-null.
+                let cmd = unsafe {
+                    OwningPtr::<Unaligned>::new(std::ptr::NonNull::new_unchecked(cursor.cast()))
+                };
+                // SAFETY: The data underneath the cursor must correspond to the type erased in metadata,
+                // since they were stored next to each other by `.push()`.
+                // For ZSTs, the type doesn't matter as long as the pointer is non-null.
+                let size = unsafe { (meta.consume_command_and_get_size)(cmd, &mut world) };
+                // Advance the cursor past the command. For ZSTs, the cursor will not move.
+                // At this point, it will either point to the next `CommandMeta`,
+                // or the cursor will be out of bounds and the loop will end.
+                // SAFETY: The address just past the command is either within the buffer,
+                // or 1 byte past the end, so this addition will not overflow the pointer's allocation.
+                cursor = unsafe { cursor.add(size) };
+
+                if let Some(world) = &mut world {
+                    // If the command we just applied generated more commands we must apply those first
+                    if !world.command_queue.is_empty() {
+                        // If our current list of commands isn't complete push it to the `resolving_commands` stack to be applied after
+                        if cursor < end {
+                            resolving_commands.push((cursor, end));
+                        }
+                        let mut bytes = std::mem::take(&mut world.command_queue.bytes);
+
+                        // Set our variables to start applying the new queue
+                        let bytes_range = bytes.as_mut_ptr_range();
+                        cursor = bytes_range.start;
+                        end = bytes_range.end;
+
+                        // Store our buffer so it is not dropped;
+                        buffers.push(bytes);
```

Comment:
At the end of the `while cursor < end {` loop, we should put this buffer back into the world (unless the world's new command queue has a bigger capacity). We should then save the smaller of the two buffers to swap back into the world in future loop iterations and reduce more allocations, as well as reducing allocations for the stack that stores the buffers.

Reply:
I think we could simplify the command queue juggling and reduce further allocations by storing a second always-empty command queue on the world, which would be used exclusively to swap out for this function and reuse allocations across calls to `CommandQueue::apply`